rust/src/tx_builder.rs

```diff
@@ -154,13 +154,94 @@ pub enum CoinSelectionStrategyCIP2 {
 
 #[wasm_bindgen]
 #[derive(Clone, Debug)]
-pub struct TransactionBuilder {
-    coins_per_utxo_word: BigNum,
-    pool_deposit: BigNum,
-    key_deposit: BigNum,
-    max_value_size: u32,
-    max_tx_size: u32,
+pub struct TransactionBuilderConfig {
     fee_algo: fees::LinearFee,
+    pool_deposit: BigNum,      // protocol parameter
+    key_deposit: BigNum,       // protocol parameter
+    max_value_size: u32,       // protocol parameter
+    max_tx_size: u32,          // protocol parameter
+    coins_per_utxo_word: Coin, // protocol parameter
+    prefer_pure_change: bool,
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Debug)]
+pub struct TransactionBuilderConfigBuilder {
+    fee_algo: Option<fees::LinearFee>,
+    pool_deposit: Option<BigNum>,      // protocol parameter
+    key_deposit: Option<BigNum>,       // protocol parameter
+    max_value_size: Option<u32>,       // protocol parameter
+    max_tx_size: Option<u32>,          // protocol parameter
+    coins_per_utxo_word: Option<Coin>, // protocol parameter
+    prefer_pure_change: bool,
+}
+
+#[wasm_bindgen]
+impl TransactionBuilderConfigBuilder {
+    pub fn new() -> Self {
+        Self {
+            fee_algo: None,
+            pool_deposit: None,
+            key_deposit: None,
+            max_value_size: None,
+            max_tx_size: None,
+            coins_per_utxo_word: None,
+            prefer_pure_change: false,
+        }
+    }
+
+    pub fn fee_algo(mut self, fee_algo: fees::LinearFee) -> Self {
+        self.fee_algo = Some(fee_algo);
+        self
+    }
+
+    pub fn pool_deposit(mut self, pool_deposit: BigNum) -> Self {
+        self.pool_deposit = Some(pool_deposit);
+        self
+    }
+
+    pub fn key_deposit(mut self, key_deposit: BigNum) -> Self {
+        self.key_deposit = Some(key_deposit);
+        self
+    }
+
+    pub fn max_value_size(mut self, max_value_size: u32) -> Self {
+        self.max_value_size = Some(max_value_size);
+        self
+    }
+
+    pub fn max_tx_size(mut self, max_tx_size: u32) -> Self {
+        self.max_tx_size = Some(max_tx_size);
+        self
+    }
+
+    pub fn coins_per_utxo_word(mut self, coins_per_utxo_word: Coin) -> Self {
+        self.coins_per_utxo_word = Some(coins_per_utxo_word);
+        self
+    }
+
+    pub fn prefer_pure_change(mut self, prefer_pure_change: bool) -> Self {
+        self.prefer_pure_change = prefer_pure_change;
+        self
+    }
+
+    pub fn build(self) -> Result<TransactionBuilderConfig, JsError> {
+        Ok(TransactionBuilderConfig {
+            fee_algo: self.fee_algo.ok_or(JsError::from_str("uninitialized field: fee_algo"))?,
```

Comment:
Is this a tradeoff we want to make? By doing this we gain descriptive setters instead of a constructor with unnamed parameters, but we lose the static checks and instead have forgetting one be a runtime error. Most editors will show you the names while typing the code so it would help more for reading code than writing it, but all the required ones are fixed (at a given time) protocol parameters anyway.

This is as opposed to having the required fields be in the constructor and only have setters for optional ones.

If you did that instead then you could also combine these two structs into one and not need a builder at all.

Reply:
A lot of libraries do this pattern in rust, clippy even gives warnings when you have too many arguments in functions. Having a seperate config type is fine I think, since we put the config struct inside the bigger struct instead of moving fields one by one. Also many people don't use editors so having too many arguments makes it more dangerous if a lot of the arguments have the same types and It causes the API to break more frequently because every time you change a argument the function API changes.

Reply:
> Did you see the comments I left in #261 regarding the config builder?

The problem choosing between type-security or runtime-security is to find the golden middle ground in this case, IMO. Main problem that we are working against is to try and protect against a faulty case when the builder can **run** with either wrong values accidentally set to wrong positions because of matching types, or some values not set at all because they have been missed. And we must consider that we are creating the lib for the JS environment as well, not just Rust, and it's being used in non-typed JS cases as well.

Two latest changes to the parameter list both were somewhat troublesome:

1. Two new parameters for `max_value_size` and `max_tx_size` been added at the very end of the list, this did not cause any problems for people in untyped JS and also did not even fail fast in runtime, because from JS to WASM those were set to default values of zero and only were failing when later calling the build function, getting people confused. Potentially if they have been added at the beginning of the list - this would cause runtime failurs for people because i32 is being passed where `LinearFee` or `BigNum` is expected, but then not only it's not that great of a solution, but it also demonstrates that you need to remember about it every single time.

2. One the existing parameters got replaced with a completely new parameter with a different meaning and value, but same exact type. This caused the problem that even the type-check would pass for people if the parameter would be on the same position in the list. This was solved by moving the parameter to the end of the list, which nicely will fail in both the type-check and the runtime in case the old parameter-list will be used, because in runtime the `BigNum` would be passed in the position where `i32` is now expected. But then the question: what would need to be done if the next parameter in the list were also a `BigNum` ? Or what will need to be done if the `coins_per_utxo_word` will get replaced again with a different-meaning parameter next time, but now they are at the end of the list ? Will we need to rearrange the parameter-list in weirder and weirder combinations every single time something like this happens ?

The solution with the config-builder is supposed to be the middle ground that solved more problems than it introduces:
1. The parameters are now visibly named and can be arranged in any order
2. If some parameter changes the name - it will fail **fast** in both the type-check (if the code uses a removed setter) and the runtime-check
3. If some new required parameters are added - it will fail **fast** in the runtime-check and plus protect from a default number value to be used silently from JS
4. It creates the visible and easy framework on how to add new parameters to the builder without having to think on how it will potentially cause problems in typed or untyped environments and thinking on whether you need to rearrange the parameters and how.

This protects the parameters from all sides and saves devs from having problems with adding new parameters. The only case in which it can be considered a reduction is when new required parameters are added - because now it will fail in runtime and not in type-check, but it is higher priority to at least fail **fast** in runtime than potentially not fail in an untyped environment at all.

Reply:
>Two new parameters for max_value_size and max_tx_size been added at the very end of the list, this did not cause any errors during tx-builder creation for people in untyped JS and also did not even fail fast in runtime, because from JS to WASM those were set to default values of zero and only were failing when later calling the build function, getting people confused.

Ah, I didn't know that happened going to wasm. That makes more sense now.